L1Trigger/TrackerTFP/src/MiniHoughTransform.cc

```diff
@@ -0,0 +1,285 @@
+        cells.reserve(numCells_);
+        const bool compA = 2. * abs(chi) < phiT_.base();
+        const bool compB = 2. * abs(chi) < abs(r * inv2R_.base());
+        const bool compAB = compA && compB;
+        if (chi >= 0. && r >= 0.) { cells.push_back(3); if (compA) cells.push_back(1); if(compAB) cells.push_back(2); }
+        if (chi >= 0. && r <  0.) { cells.push_back(1); if (compA) cells.push_back(3); if(compAB) cells.push_back(0); }
+        if (chi <  0. && r >= 0.) { cells.push_back(0); if (compA) cells.push_back(2); if(compAB) cells.push_back(1); }
+        if (chi <  0. && r <  0.) { cells.push_back(2); if (compA) cells.push_back(0); if(compAB) cells.push_back(3); }
+        // organise stubs in finer track candidates
+        for (int cell : cells) {
+          const int inv2R = cell / setup_->mhtNumBinsPhiT();
+          const int phiT = cell % setup_->mhtNumBinsPhiT();
+          stubsMHT_.emplace_back(**stub, phiT, inv2R);
+          mhtCells[cell].push_back(&stubsMHT_.back());
+        }
+      }
+      // perform pattern recognition
+      for (int sel = 0; sel < numCells_; sel++) {
+        deque<StubMHT*>& stream = streams[channel * numCells_ + sel];
+        vector<StubMHT*>& mhtCell = mhtCells[sel];
+        set<int> layers;
+        auto toLayer = [](StubMHT* stub){ return stub->layer(); };
+        transform(mhtCell.begin(), mhtCell.end(), inserter(layers, layers.begin()), toLayer);
+        if ((int)layers.size() < setup_->mhtMinLayers())
+          mhtCell.clear();
+        for (StubMHT* stub : mhtCell)
+          stream.push_back(stub);
+        stream.insert(stream.end(), size - (int)mhtCell.size(), nullptr);
+      }
+    }
+    for (int sel = 0; sel < numCells_; sel++) {
+      deque<StubMHT*>& stream = streams[channel * numCells_ + sel];
+      // remove all gaps between end and last stub
+      for(auto it = stream.end(); it != stream.begin();)
+        it = (*--it) ? stream.begin() : stream.erase(it);
+      // read out fine track cannot start before rough track has read in completely, add gaps to take this into account
+      int pos(0);
+      for (auto it = stream.begin(); it != stream.end();) {
+        if (!(*it)) {
+          it = stream.erase(it);
+          continue;
+        }
+        id = (*it)->trackId();
+        const int s = distance(it, find_if(it, stream.end(), differentMHT));
+        const int d = distance(stream.begin(), it);
+        pos += s;
+        if (d < pos) {
+          const int diff = pos - d;
+          it = stream.insert(it, diff, nullptr);
+          it = next(it, diff);
+        }
+        else
+          it = stream.erase(remove(next(stream.begin(), pos), it, nullptr), it);
+        it = next(it, s);
+      }
+      // adjust stream start so that first output stub is in first place in case of quickest track
+      if (!stream.empty())
+        stream.erase(stream.begin(), next(stream.begin(), setup_->mhtMinLayers()));
+    }
+  }
+
+  // mux 4 streams to 1 stream
```

Comment:
Change comment to "Static load balancing of inputs: mux 4 streams to 1 stream". Ideally add 1 sentence also to say which streams are muxed.

Reply:
done